### PR TITLE
Allow specifying cipher suites instead of single ciphers

### PR DESCRIPTION
**Pull request by quantum5**
Changes proposed in this pull request:
* `CipherName` now specifies all cipher suites instead of the preferred cipher. This allows insecure ciphers like RC4 to be permanently disabled, instead of being the default fallback when the preferred cipher is unsupported.
* `CipherName` is now left for OpenSSL to verify. Should it be invalid, a secure default is used. The default `CipherName` setting for new servers is one such invalid string: `~DEFAULT~`. This allows for future updates to change the default and the servers can stay secure.
* This allows getting a respectable A on the SSL Labs test. (There is no point to enabling HSTS for a VPN, for example.)

Option 1 is chosen.

### CODE REVIEW

File: src/Mayaqua/Network.c [hunk 1]
```diff
@@ -12785,12 +12764,7 @@ void SetWantToUseCipher(SOCK *sock, char *name)
 		Free(sock->WaitToUseCipher);
 	}
 
-	Zero(tmp, sizeof(tmp));
-	StrCpy(tmp, sizeof(tmp), name);
-	StrCat(tmp, sizeof(tmp), " ");
-	StrCat(tmp, sizeof(tmp), cipher_list);
```
Comment (moatazelmasry2) [inline src/Mayaqua/Network.c]: If tmp is not being used anywhere, then its definition should also be deleted

File: src/Mayaqua/Network.h [hunk 2]
```diff
@@ -153,6 +153,7 @@ struct DYN_VALUE
 #define	DEFAULT_GETIP_THREAD_MAX_NUM		64
 #endif	// USE_STRATEGY_LOW_MEMORY
 
+#define	DEFAULT_CIPHER_LIST			"ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20:ECDHE+AES256:DHE+AES256:RSA+AES"
```
Comment (moatazelmasry2) [inline src/Mayaqua/Network.h]: this list is quite different from cipher_list and I wounder how/why you choose those ciphers and not others used in cipher_list??

**Comment by moatazelmasry2** [comment]
what about cipher_list defined in 
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/Mayaqua/Network.c#L236
?

In practice it is disabled, but it is still there and being used by a couple of functions.

**Comment by moatazelmasry2** [comment]
This is a quite important PR btw so thank you for that and when merged, it will solve the following issues:
Issue #147
Issue #179 
Issue #253
Issue #264

**Comment by quantum5** [comment]
@moatazelmasry2 the list of ciphers are chosen to be in order of security, as typically recommended, preferring authenticated encryption ciphers, while also maintaining compatibility. The intention is that most installations will use these default ciphers and automatically benefit from the increased security should new ciphers be created.

Using the other cipher list would prefer using extremely insecure ciphers like RC4-MD5. I would remove it, but the management tool uses it and I couldn't quite figure out how to modify it.

File: src/Mayaqua/Network.c (hunk 1 above)
Comment (moatazelmasry2) [reply src/Mayaqua/Network.c]: Please remove the tmp variable

**Comment by moatazelmasry2** [comment]
ok. Thanks for the explaination and I agree. But I'm not really comfortable with giving invalid value to the default cipher and enforce SE to get something from SSL. It feels like a hack. Can we maybe define the cipher suites directly in the cipher_list variable to achieve the same result?

**Comment by quantum5** [comment]
The main concern regarding configurable ciphers is what happens if the server is misconfigured to use no cipher at all. I decided to use the fallback as a sane alternative, since otherwise a lot of error handline will have to be implemented so some obscure error message doesn't show up.

Admittedly, the invalid specification is a hack, but is probably the best I can do without seriously messing with the management UI. The management program cuts off any long cipher list, which makes it almost impossible to list more than one cipher inside it, hence, a short invalid string has to be used for now.

To do it properly, the UI will have to be a list of checkboxes that can be moved around, or perhaps two lists of ciphers, one for active, and one for inactive, and a button to move ciphers in and out of the list. I am afraid I do not have the time for this.

**Comment by quantum5** [comment]
Also note that OpenSSL will use a default list of ciphers anyway, if it fails to set the cipher list.

**Comment by moatazelmasry2** [comment]
I really need a second opinion on this. @dnobori @SoftEtherVPN/softether-development-board-members  could you please review this PR and especially whether the introduced hack migh cause any vulnerability

**Comment by moatazelmasry2** [comment]
I merged this finally into master. This looks to be working fine and an expected cipher is being choosen when connecting. 

Thank you for your hard work.

**Comment by the6thBook** [comment]
I am still having an issue (Version 4.29 Build 9680   ).  I am still showing RC4-MD5 via nessus vulnerability scan and openssl sclient -connect x.x.x. -cipher RC4-MD5, even though, I have it disabled:

VPN Server/DEFAULT>ServerCipherGet
ServerCipherGet command - Get the Encrypted Algorithm Used for VPN Communication.
Encrypted Algorithm Currently Used by VPN Server: 
 AES256-SHA

List of Usable Encrypted Algorithm Names
...

Also have set:
		bool AcceptOnlyTls true
		bool Tls_Disable1_0 true
		bool Tls_Disable1_1 true
		bool Tls_Disable1_2 false


When I change AcceptOnlyTls to false, it seems like it gets rid of md5, but then sslv3 is enabled.

**Comment by davidebeatrici** [comment]
The stable version still has SSLv3 support for backwards compatibility, however the options in the configuration don't seem to work as expected.

If you want to get rid of SSLv3 completely, please build SoftEther VPN from this repository.

**Comment by dnobori** [comment]
When AcceptOnlyTls to false:
```
$ openssl s_client -connect 127.0.0.1 -cipher RC4-MD5 -ssl3

New, TLSv1/SSLv3, Cipher is RC4-MD5
Server public key is 2048 bit
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : SSLv3
    Cipher    : RC4-MD5
    Session-ID: 4F14D4AF1B458C93F6A0762574BB263DD6D61633C138A53FFCBBBA3B9C05E773
    Session-ID-ctx:
    Master-Key: 4CDE19B4EE3934D6E7755BBB94FCFFC5AF25692DEFFD506FBD0AE3908EC80ADB21502374458FAF937527B905F79F7A45
    Key-Arg   : None
    Start Time: 1563242138
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
```

When AcceptOnlyTls to true:
```
$ openssl s_client -connect 127.0.0.1 -cipher RC4-MD5 -ssl3
CONNECTED(00000003)
15772:error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure:s3_pkt.c:1053:SSL alert number 40
15772:error:1409E0E5:SSL routines:SSL3_WRITE_BYTES:ssl handshake failure:s3_pkt.c:530:
```

My results look very normal.

**Comment by the6thBook** [comment]
> When AcceptOnlyTls to false:
> 
> ```
> $ openssl s_client -connect 127.0.0.1:443 -cipher RC4-MD5 -ssl3
> 
> New, TLSv1/SSLv3, Cipher is RC4-MD5
> Server public key is 2048 bit
> Compression: NONE
> Expansion: NONE
> SSL-Session:
>     Protocol  : SSLv3
>     Cipher    : RC4-MD5
>     Session-ID: 4F14D4AF1B458C93F6A0762574BB263DD6D61633C138A53FFCBBBA3B9C05E773
>     Session-ID-ctx:
>     Master-Key: 4CDE19B4EE3934D6E7755BBB94FCFFC5AF25692DEFFD506FBD0AE3908EC80ADB21502374458FAF937527B905F79F7A45
>     Key-Arg   : None
>     Start Time: 1563242138
>     Timeout   : 7200 (sec)
>     Verify return code: 18 (self signed certificate)
> ```
> 
> When AcceptOnlyTls to true:
> 
> ```
> $ openssl s_client -connect 127.0.0.1:443 -cipher RC4-MD5 -ssl3
> CONNECTED(00000003)
> 15772:error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure:s3_pkt.c:1053:SSL alert number 40
> 15772:error:1409E0E5:SSL routines:SSL3_WRITE_BYTES:ssl handshake failure:s3_pkt.c:530:
> ```
> 
> My results look very normal.

Set your Cipher to AES256-SHA... Do you still get RC4-MD5?  That is my issue.

**Comment by dnobori** [comment]
I tested with
```
		bool AcceptOnlyTls false
		string CipherName AES256-SHA
```
and run
```
$ openssl s_client -connect 127.0.0.1:443 -cipher RC4-MD5 -ssl3
```
results
```
New, TLSv1/SSLv3, Cipher is RC4-MD5
Server public key is 2048 bit
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : SSLv3
    Cipher    : RC4-MD5
    Session-ID: 8CB669DE1450CE53257D721FDBF6B29FC3899B450024F4C62C9414479B4D5E78
    Session-ID-ctx:
    Master-Key: 36A25554780EA903834D138106D7E0D6EE04561D70F908BC4C0CD0662CBD10FCC56FF8EF67E4FF44229243ACC087707A
    Key-Arg   : None
    Start Time: 1563325230
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
```

It seems there are no problem in my environment.

**Comment by the6thBook** [comment]
I'm still seeing RC4-MD5.  

VPN Server/DEFAULT>ServerCipherGet
ServerCipherGet command - Get the Encrypted Algorithm Used for VPN Communication.
Encrypted Algorithm Currently Used by VPN Server:
AES256-SHA

List of Usable Encrypted Algorithm Names
...

Also have set:
bool AcceptOnlyTls true
bool Tls_Disable1_0 true
bool Tls_Disable1_1 true
bool Tls_Disable1_2 false


New, TLSv1/SSLv3, Cipher is RC4-MD5
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : RC4-MD5
    Session-ID: 6906F7579CBB8231D0D2536977A6775FB37FED79400B4BBC0075A39D77A8E62F
    Session-ID-ctx: 
    Master-Key: 39A810594AF34650CCACFDB1ED0EFFA424B33789C9A23611098590365C613DD090EBD83084B0AEFCDC5BF06899016E76
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1563460705
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)

Anything I should try?

**Comment by dnobori** [comment]
Perhaps TLS 1.2 doesn't support AES256-SHA.
Try to set
```
DHE-RSA-AES256-SHA, AES256-GCM-SHA256, DHE-RSA-AES256-GCM-SHA256 or ECDHE-RSA-AES256-GCM-SHA256
```

**Comment by the6thBook** [comment]
I still have this issue with RC4-MD5, I've tried:
DHE-RSA-AES256-SHA
AES256-GCM-SHA384
DHE-RSA-AES128-GCM-SHA256

The correct cipher show up as preferred in sslscan, but still can connect with RC4-MD5 with openssl s_client and the RC4-MD5 shows up on Nessus vulnerability scanner as plugin 65821 - "SSL RC4 Cipher Suites Supported (Bar Mitzvah)", CVE-2013-2566, CVE-2015-2808.

**Comment by the6thBook** [comment]
Any updates on this?

**Comment by malikshi** [comment]
`bool AcceptOnlyTls false`
is it still valid for latest build from this repo?

**Comment by davidebeatrici** [comment]
It's implicit, SSL is not permitted anymore.

**Comment by malikshi** [comment]
thank you. gonna use the RTM version then, the old router can't connect

**Comment by the6thBook** [comment]
This is still an issue in:
Version 4.41 Build 9787   (English)
Compiled 2023/03/14 11:00:03 by buildsan at crosswin with OpenSSL 3.0.7

I set the cipher suite:
_string CipherName AES256-GCM-SHA384_

 but I still can connect with RC4-SHA.
openssl s_client -connect vpn:443 -cipher RC4-SHA
...
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : RC4-SHA
....

**Pull request merged**